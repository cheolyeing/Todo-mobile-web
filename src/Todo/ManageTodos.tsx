import Todo from "./Todo";

class TodoManager {
  public todos: Todo[];
  private TODO_LS: string = "todos";

  constructor() {
    this.todos = [];
  }

  loadTodos = () => {
    const loadedTodos = localStorage.getItem(this.TODO_LS);
    if (loadedTodos === null) {
    } else {
      const parsedTodos = JSON.parse(loadedTodos);
      this.todos = parsedTodos;
    }
  };

  saveTodos = (sort?: boolean) => {
    if (sort) this.sortById();
    localStorage.setItem(this.TODO_LS, JSON.stringify(this.todos));
  };

  addTodo = (newTodo: Todo) => {
    this.todos = [...this.todos, newTodo];
    this.saveTodos();
  };

  updateTodo = (updatedTodo: Todo) => {
    const idx = this.todos.findIndex((todo) => {
      return todo.id === updatedTodo.id;
    });
    if (idx > -1) {
      this.todos[idx].title = updatedTodo.title;
      this.todos[idx].content = updatedTodo.content;
      this.todos[idx].dueDate = updatedTodo.dueDate;
      this.todos[idx].done = updatedTodo.done;
    }
    this.saveTodos();
  };

  getNewId = () => {
    if (this.todos.length === 0) return 0;
    return this.todos[this.todos.length - 1].id + 1;
  };

  sortById = () => {
    this.todos.sort(function (a, b) {
      return a.id - b.id;
    });
  };

  deleteById = (id: number) => {
    const idx = this.todos.findIndex((todo) => {
      return todo.id === id;
    });
    if (idx > -1) this.todos.splice(idx, 1);
    this.saveTodos();
  };

  changeDoneById = (id: number) => {
    const idx = this.todos.findIndex((todos) => {
      return todos.id === id;
    });
    if (idx > -1) this.todos[idx].done = !this.todos[idx].done;
    this.saveTodos();
  };
}

export default TodoManager;
