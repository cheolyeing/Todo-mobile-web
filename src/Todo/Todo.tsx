class Todo {
  id: number;
  title: string;
  content: string;
  done: boolean;
  dueDate?: string;

  constructor(
    id: number,
    title: string,
    content: string,
    done: boolean,
    dueDate?: string
  ) {
    this.id = id;
    this.title = title;
    this.content = content;
    this.done = done;
    this.dueDate = dueDate;
  }

  setDone() {
    this.done = !this.done;
  }
}

export default Todo;
