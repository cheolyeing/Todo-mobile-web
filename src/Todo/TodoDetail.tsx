import React, { useEffect, useState } from "react";

import TodoManager from "./TodoManager";

import {
  Title,
  TodoDeleteBtn,
  TodoContainer,
  TodoTitle,
  TodoDeleteBtnContainer,
  TodoDueDate,
  TodoInside,
  TodoCheckBox,
  TodoCheckBoxContainer,
} from "../Style";

function TodoDetail(props: any) {
  const DUE_DATE = "due date: ";
  const id: number = props.todo.id;
  const title: string = props.todo.title;
  const dueDate: string = props.todo.dueDate;
  const [done, setDone] = useState<boolean>(props.todo.done);
  const [dateOver, setDateOver] = useState<boolean>(false);
  const todoManager: TodoManager = new TodoManager();
  const e = props.todo;
  const theme = {
    done: done,
    dueDate: dueDate !== undefined && dueDate !== null && dueDate !== "",
    dateOver: dateOver,
  };

  const checkDateOver = () => {
    let checkdDateOver: boolean = false;

    if (dueDate === undefined || dueDate === null || dueDate === "") {
    } else {
      let today: Date = new Date();
      let split: string[] = dueDate.split("-");
      if (+split[0] < today.getFullYear()) checkdDateOver = true;
      if (+split[1] < today.getMonth() + 1) checkdDateOver = true;
      if (+split[2] <= today.getDate()) checkdDateOver = true;
    }
    setDateOver(checkdDateOver);
  };

  const handleClickDone = () => {
    todoManager.loadTodos();
    todoManager.changeDoneById(id);
    setDone(!done);
    props.reload();
  };

  const handleClickDelete = () => {
    todoManager.loadTodos();
    todoManager.deleteById(id);
    props.reload();
  };

  const showDuedate = () => {
    if (dueDate !== undefined && dueDate !== null && dueDate !== "") {
      return (
        <TodoDueDate theme={theme}>
          {DUE_DATE}
          {dueDate}
        </TodoDueDate>
      );
    }
  };
  useEffect(() => {
    checkDateOver();
  }, []);
  return (
    <TodoContainer>
      <TodoCheckBoxContainer theme={theme}>
        <TodoCheckBox
          type="checkbox"
          checked={done}
          onChange={handleClickDone}
        />
      </TodoCheckBoxContainer>

      <TodoInside>
        <Title
          to={{
            pathname: `/Todo-mobile-web/editTodo/${id}`,
            state: {
              e,
            },
          }}
        >
          {" "}
          <div>
            <TodoTitle theme={theme}>{title}</TodoTitle>
            {showDuedate()}
          </div>
        </Title>
      </TodoInside>

      <TodoDeleteBtnContainer theme={theme}>
        <TodoDeleteBtn onClick={handleClickDelete}>
          <span role="img" aria-label="x">
            ‚ùå
          </span>
        </TodoDeleteBtn>
      </TodoDeleteBtnContainer>

      <hr />
    </TodoContainer>
  );
}

export default TodoDetail;
