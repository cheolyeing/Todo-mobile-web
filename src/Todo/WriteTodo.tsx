import React, { useState } from "react";

import { Input, ContentArea, SaveLink, SaveDiv, Header3 } from "../Style";

import Todo from "./Todo";
import TodoManager from "./TodoManager";

function WriteTodo(props: any) {
  const TODO: string = "To-Do";
  const TITLE: string = "제목";
  const CONTENT: string = "내용";
  const DUE_DATE: string = "Due Date (Option)";
  const SAVE_BUTTON: string = "저장 버튼";

  const todo = props.location.state.e;
  let id = todo.id;
  const [title, setTitle] = useState<string>(todo.title);
  const [content, setContent] = useState<string>(todo.content);
  const [dueDate, setDueDate] = useState<string>(todo.dueDate);
  const done = todo.done;

  const onClickSaveBtn = (error: any) => {
    if (!titleCheck()) {
      alert("제목을 입력해주세요!");
      error.preventDefault();
      return;
    }
    const todoManager: TodoManager = new TodoManager();
    todoManager.loadTodos();

    if (id < 0) {
      todoManager.addTodo(
        new Todo(todoManager.getNewId(), title, content, false, dueDate)
      );
    } else {
      todoManager.updateTodo(new Todo(id, title, content, done, dueDate));
    }
  };

  const titleCheck = () => {
    if (title === "") {
      return false;
    }
    return true;
  };

  return (
    <div>
      <div>
        <h1>{TODO}</h1>
        <Header3>{TITLE}</Header3>
        <SaveDiv>
          <ContentArea
            rows={1}
            value={title}
            onChange={(e) => {
              setTitle(e.target.value);
            }}
            placeholder="텍스트 입력"
          />
        </SaveDiv>
        <Header3>{CONTENT}</Header3>
        <SaveDiv>
          <ContentArea
            rows={10}
            value={content}
            onChange={(e) => {
              setContent(e.target.value);
            }}
            placeholder="텍스트 입력"
          />
        </SaveDiv>
        <Header3>{DUE_DATE}</Header3>
        <SaveDiv>
          <Input
            name="duedate"
            type="date"
            value={dueDate}
            onChange={(e) => {
              setDueDate(e.target.value);
            }}
          />
        </SaveDiv>
      </div>
      <SaveLink to="/Todo-mobile-web">
        <SaveDiv
          onClick={(e) => {
            onClickSaveBtn(e);
          }}
        >
          {SAVE_BUTTON}
        </SaveDiv>
      </SaveLink>
    </div>
  );
}

export default WriteTodo;
