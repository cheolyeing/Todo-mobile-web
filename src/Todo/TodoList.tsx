import React, { useEffect, useState } from "react";

import {
  TodoListContainer,
  AddButton,
  Header3,
  AddLink,
  TodoContainer,
} from "../Style";

import Todo from "./Todo";
import TodoManager from "./ManageTodos";
import TodoDetail from "./TodoDetail";

function TodoList() {
  const TITLE: string = "이번주 To-Do";
  const MAKE_TODO: string = "추가 버튼";
  const [todos, setTodos] = useState<Todo[]>([]);
  const e = new Todo(-1, "", "", false);

  const init = () => {
    const todoManager: TodoManager = new TodoManager();
    todoManager.loadTodos();
    setTodos(todoManager.todos);
  };

  useEffect(() => {
    init();
  }, []);

  return (
    <div>
      <div>
        <Header3>{TITLE}</Header3>
      </div>
      <div>
        <AddLink
          to={{
            pathname: `/newTodo/${e.id}`,
            state: {
              e,
            },
          }}
        >
          <AddButton>{MAKE_TODO}</AddButton>
        </AddLink>
      </div>
      <br />
      <TodoListContainer>
        <TodoContainer />
        {todos.map((e) => {
          return <TodoDetail key={e.id} todo={e} reload={init} />;
        })}
      </TodoListContainer>
    </div>
  );
}

export default TodoList;
