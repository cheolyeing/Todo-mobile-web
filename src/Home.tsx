import React from "react";
import Title from "./Title";
import WeeklyWeeather from "./Weather/WeeklyWeather";
import TodoList from "./Todo/TodoList";

function Home() {
  return (
    <div>
      <Title />
      <WeeklyWeeather />
      <TodoList />
    </div>
  );
}

export default Home;
