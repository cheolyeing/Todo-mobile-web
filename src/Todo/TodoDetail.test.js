import React from "react";
import ReactDOM from "react-dom";
import { render, renderer } from "@testing-library/react";
import { BrowserRouter, Switch, Route } from "react-router-dom";

import TodoDetail from "./TodoDetail";
import Todo from "./Todo";
import Home from "../Home";
import WriteTodo from "./WriteTodo";

describe("<TodoDetail />", () => {
  it("matches snapshot", () => {
    const utils = render(
      <BrowserRouter>
        <TodoDetail
          todo={
            new Todo(
              123,
              "Todo Title",
              "What I have to do",
              false,
              "2020-09-01"
            )
          }
        />
      </BrowserRouter>
    );
    expect(utils.container).toMatchSnapshot();
  });
  it("shows the props correctly", () => {
    const utils = render(
      <BrowserRouter>
        <TodoDetail
          todo={
            new Todo(
              123,
              "Todo Title",
              "What I have to do",
              false,
              "2020-09-01"
            )
          }
        />
      </BrowserRouter>
    );
    utils.getByText("Todo Title");
    utils.getByText("‚ùå");
  });
});
