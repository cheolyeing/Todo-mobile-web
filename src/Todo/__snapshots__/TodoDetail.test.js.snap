// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoDetail /> matches snapshot 1`] = `
<div>
  <div
    class="sc-iqAbSa ZilSU"
  >
    <div
      class="sc-fFSRdu bYHtHB"
    >
      <input
        class="sc-fnVYJo TqA-dN"
        type="checkbox"
      />
    </div>
    <div
      class="sc-crzoUp lkusge"
    >
      <a
        class="sc-bdnylx hjKUCK"
        href="/Todo-mobile-web/editTodo/123"
      >
         
        <div>
          <div
            class="sc-dIsAE IMwkE"
          >
            Todo Title
          </div>
          <div
            class="sc-bqGHjH eOlJBZ"
          >
            due date: 
            2020-09-01
          </div>
        </div>
      </a>
    </div>
    <div
      class="sc-ksluoS iQwSjW"
    >
      <button
        class="sc-hBMVcZ jTHzhK"
      >
        <span
          aria-label="x"
          role="img"
        >
          ‚ùå
        </span>
      </button>
    </div>
    <hr />
  </div>
</div>
`;
